File: services/api/app/models/flashcard.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime, timedelta

class Flashcard(BaseModel):
    id: Optional[str] = None
    subject_id: str
    file_id: Optional[str] = None
    front: str
    back: str
    
    # SM-2 Parameters
    repetition: int = 0
    interval: int = 0
    ease_factor: float = 2.5
    next_review: datetime = Field(default_factory=datetime.now)
    
    # Status
    status: str = "new" # new, learning, mastered (derived from interval/repetition)
    
    class Config:
        populate_by_name = True
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

class FlashcardGenerationRequest(BaseModel):
    subject_id: str
    file_id: Optional[str] = None
    text_content: str
    count: int = 10

class CreateCardRequest(BaseModel):
    subject_id: str
    file_id: Optional[str] = None
    front: str
    back: str

class ReviewRequest(BaseModel):
    card_id: str
    rating: int # 1 (Need Review) - 5 (Easy/Got it)

File: apps/mobile/lib/features/flashcards/domain/flashcard.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class Flashcard {
  final String id;
  final String subjectId;
  final String? fileId;
  final String front;
  final String back;
  final String status; // 'new', 'learning', 'mastered'
  final DateTime nextReview;
  final int repetition;
  final int interval;
  final double easeFactor;

  Flashcard({
    required this.id,
    required this.subjectId,
    this.fileId,
    required this.front,
    required this.back,
    required this.status,
    required this.nextReview,
    required this.repetition,
    required this.interval,
    required this.easeFactor,
  });

  factory Flashcard.fromJson(Map<String, dynamic> json, String id) {
    return Flashcard(
      id: id,
      subjectId: json['subject_id'] ?? '',
      fileId: json['file_id'],
      front: json['front'] ?? '',
      back: json['back'] ?? '',
      status: json['status'] ?? 'new',
      nextReview: _parseDate(json['next_review']),
      repetition: json['repetition'] ?? 0,
      interval: json['interval'] ?? 0,
      easeFactor: (json['ease_factor'] ?? 2.5).toDouble(),
    );
  }

  static DateTime _parseDate(dynamic value) {
    if (value == null) return DateTime.now();
    if (value is Timestamp) return value.toDate();
    if (value is String) return DateTime.parse(value);
    return DateTime.now();
  }
}
